File: st-util/src/main/java/de/saring/util/data/IdObject.java
package de.saring.util.data;

import java.util.Objects;

/**
 * Abstract base class for all objects which needs to have an ID for
 * referencing.
 *
 * @author Stefan Saring
 * @version 1.0
 */
public abstract class IdObject {

    /**
     * The ID of the object for referencing.
     */
    private int id;

    /**
     * Standard c'tor.
     *
     * @param id the ID of the object
     */
    public IdObject(int id) {
        this.id = id;
    }

    public int getId() {
        return id;
    }

    /**
     * Compares the specified object with this object. The objects are equal
     * when they are of the same type and when they have the same ID.
     *
     * @param obj the object to compare with
     * @return true when same type and ID, false otherwise
     */
    @Override
    public final boolean equals(Object obj) {
        if (obj == null || this.getClass() != obj.getClass()) {
            return false;
        } else if (this == obj) {
            return true;
        } else {
            IdObject other = (IdObject) obj;
            return this.id == other.id;
        }
    }

    @Override
    public final int hashCode() {
        return Objects.hashCode(this.id);
    }

    @Override
    public String toString() {
        StringBuilder sBuilder = new StringBuilder();
        sBuilder.append(this.getClass().getName()).append(":\n");
        sBuilder.append(" [id=").append(this.id).append("]\n");
        return sBuilder.toString();
    }
}

File: st-util/src/main/kotlin/de/saring/util/data/IdObject.kt
package de.saring.util.data

import java.util.Objects

/**
 * Abstract base class for all objects which needs to have an ID for referencing. The ID can't be changed.
 *
 * @property id the ID of the object (null for new, not persisted objects)
 *
 * @author Stefan Saring
 */
abstract class IdObject(val id: Long?) {

    /**
     * Compares the specified object with this object. The objects are equal when they are of the same instance or when
     * of same type and they have the same ID.
     *
     * @param other the object to compare with
     * @return true when same type and ID, false otherwise
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        return when(other) {
            null -> false
            is IdObject -> {
                if (this.javaClass == other.javaClass) {
                    return if (this.id == null && other.id == null) {
                        this === other
                    } else {
                        this.id == other.id
                    }
                }
                return false
            }
            else -> this === other
        }
    }

    override fun hashCode(): Int = Objects.hashCode(this.id)

    override fun toString(): String = "${this.javaClass.name}: id=$id"
}
